Add a storage abstraction

bk2204 opened the pull request:
Currently we have a storer abstraction, but it isn't exposed, making it hard to mock the disk in unit tests that use this module.  This series wraps both the existing storers and the pack piece in a storage abstraction that can be overridden and used elsewhere.  It also provides a consistent error type so that we can tell when an object is missing versus when there's some more serious error.

This series is marked RFC because while I think it should be done, I'm finalizing the use of it in git-lfs/git-lfs and I want to be sure that if I need to make any changes that I do it before this series merges.

bk2204 commented:
I don't think I'll need to make any further changes to this outside of review improvements, so I've removed the RFC tag from this.

ttaylorr commented on errors/errors.go:
```diff
@@ -0,0 +1,22 @@
+
+func IsNoSuchObject(e error) bool {
+	_, ok := e.(*NoSuchObject)
+	return ok
```
One thing we might consider is also checking that `e != nil`, since `(*NoSuchObject)(nil)` (read: "cast `nil` to type `*NoSuchObject`)" is a valid thing to accept in this function.

bk2204 replied:
Due to nils of different types comparing unequal, I would have to do something like this:

```go
	err, ok := e.(*noSuchObject)
	return ok && err != nil
```

I've squashed that in.

ttaylorr commented on errors/errors.go:
```diff
@@ -0,0 +1,22 @@
+
+// NoSuchObject is an error type that occurs when no object with a given object
+// ID is available.
+type NoSuchObject struct {
```
Another thing we might think about is making this name `NoSuchObject` the constructor for an otherwise-private implementation, which I think would read nicely when returning an error.

For example, instead of:

```go
return nil, &errors.NoSuchObject{Oid: oid}
```

we could write:

```go
return nil, errors.NoSuchObject(oid)
```

ttaylorr commented on pack/set.go:
```diff
@@ -181,5 +183,5 @@ func (s *Set) each(name []byte, fn iterFn) (*Object, error) {
 	}
 
-	return nil, errNotFound
+	return nil, &errors.NoSuchObject{Oid: name}
```
✨ ✨ ✨

ttaylorr commented on backend.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+// NewFilesystemBackend initializes a new filesystem-based backend.
+func NewFilesystemBackend(root, tmp string) (storage.Backend, error) {
```
Nit: what do you think about breaking the related constructors and types into their own files in package `storage`.

bk2204 replied:
I wanted to do that, but we can't, because the `newFileStorer` and `newMemoryStorer` functions are private.  I'll update the commit message to reflect that.

ttaylorr commented on object_db.go:
```diff
@@ -256,52 +269,33 @@ func (d *ObjectDatabase) encodeBuffer(object Object, buf io.ReadWriter) (sha []b
-			return nil, err
+		if s.IsCompressed() {
+			return NewObjectReadCloser(f)
+		} else {
```
Nit: some linters will complain that an `else` block that follows an `if` with an unconditional return is redundant, and should be rewritten as:

```go
if foo {
        return bar
}
return baz
```

bk2204 replied:
Yeah, thanks for pointing this out.  I always do this in C, but for some reason I haven't yet convinced myself to do it in Go.

bk2204 commented on storage/backend.go:
```diff
@@ -0,0 +1,10 @@
+	// Storage returns a list of read sources and optionally one write
+	// source.  Generally, the write location, if present, should also be a
+	// read location.
+	Storage() ([]ReadOnlyStorage, Storage)
```
I like that idea.  Thanks for the suggestion.

bk2204 commented on storage/storage.go:
```diff
@@ -0,0 +1,30 @@
+// Storage implements an interface for reading and writing objects in an
+// object database.
+type Storage interface {
+	ReadOnlyStorage
```
Okay, sounds good.

bk2204 commented on pack/storage.go:
```diff
@@ -0,0 +1,70 @@
+	return d.mr.Read(b)
+}
+
+func (d *delayedObjectReadWriter) Write(b []byte) (int, error) {
```
This is because `Open` returns an `io.ReadWriteCloser` and we're reusing the `Open` method from the `storer` interface.  I'm not completely clear on why we need an `io.ReadWriteCloser` instead of an `io.ReadCloser`, but let me spend some more time looking at it to see if I can adjust it to use `io.ReadCloser` instead.

bk2204 commented:
I think I've addressed all of the issues mentioned here, either through code changes or commit message updates.  Do note that there are some additional commits in order to implement the suggestions made.
